Add <input type=checkbox switch> internal pseudo-elements

annevk opened the pull request:
#### dc51e7fcb906549a53e52f7c08dbb399f134334e
<pre>
Add &lt;input type=checkbox switch&gt; internal pseudo-elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=263928">https://bugs.webkit.org/show_bug.cgi?id=263928</a>
<a href="https://rdar.apple.com/117708673">rdar://117708673</a>

Reviewed by Aditya Keerthi and Antti Koivisto.

Changing the internal appearance of internal shadow tree elements in
resolveCustomStyle() does not work completely as this is too late for
Adjuster::adjust() to consider it having an appearance for the purposes
of theming. In particular it does not allow us to set intrinsic sizes
for these elements.

As such, internal shadow tree elements need to have appearance:inherit
declared upon them upfront in a user agent style sheet.

As we are planning on eventually supporting pseudo-elements for the
track and thumb of switches, add those now, but only expose them
internally until we are sure they match the eventual standard. These
pseudo-elements can then be used to set appearance:inherit in the user
agent style sheet on the internal shadow tree elements.

This appearance value (when it computes to auto) means that
autoAppearanceForElement() is invoked at which point we can then map
them to the appropriate internal values for the track and thumb.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElementType):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorPseudoElementTypeMap.in:
* Source/WebCore/css/htmlSwitchControl.css:
(input[type=&quot;checkbox&quot;][switch]::thumb, input[type=&quot;checkbox&quot;][switch]::track):
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::parsePseudoElementSelector):
* Source/WebCore/css/parser/CSSParserSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::thumb):
(WebCore::ShadowPseudoIds::track):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:
* Source/WebCore/html/shadow/SwitchThumbElement.cpp:
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::resolveCustomStyle): Deleted.
* Source/WebCore/html/shadow/SwitchThumbElement.h:
* Source/WebCore/html/shadow/SwitchTrackElement.cpp:
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::resolveCustomStyle): Deleted.
* Source/WebCore/html/shadow/SwitchTrackElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):

Canonical link: <a href="https://commits.webkit.org/270037@main">https://commits.webkit.org/270037@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff46ecb88c6f746234864745a7587b6661f9e45d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26494 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/22418 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4110 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/29 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24608 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/1980 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/21047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27080 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21973 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/22205 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/22283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1653 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/29 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5833 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/2061 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2021 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/SelectorPseudoElementTypeMap.in:
```diff
@@ -22,3 +22,5 @@ placeholder, PseudoElementWebKitCustom
 selection
 slotted
+thumb, PseudoElementWebKitCustom
+track, PseudoElementWebKitCustom
```
I would be happy to re-order this file lexicographically. Seems a bit random at the moment.

annevk commented on Source/WebCore/css/htmlSwitchControl.css:
```diff
@@ -25,5 +25,10 @@
 @namespace "http://www.w3.org/1999/xhtml";
 
 input[type="checkbox"][switch] {
+    appearance: auto;
```
This might not be needed as it would already get this from `input[type="checkbox"]` I think.

pxlcoder replied:
Yeah, I think we can remove this.

annevk commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -493,6 +494,17 @@ StyleAppearance RenderTheme::autoAppearanceForElement(RenderStyle& style, const
             return StyleAppearance::InnerSpinButton;
+
+        if (auto host = element->shadowHost()) {
+            if (host->computedStyle()->effectiveAppearance() != StyleAppearance::Auto)
```
I tested this and this works well, but I'm not familiar with these internal APIs and whether it's okay to invoke them here.

pxlcoder replied:
`autoAppearanceForElement` should only return the `auto` appearance value. If we want to drop it due to the host having `appearance: none`, `adjustAppearanceForElement` would be a better place to do it.

But... I wonder if we can `appearance: inherit !important` in our style sheet instead? Not sure if that will work, but I think it's worth trying.

nt1m replied:
I wouldn't use `!important`, but inherit could be a good idea. It would need some care if there's a complex shadow structure though.

annevk replied:
I think `!important` makes sense as it ensures you cannot get half of the UA styling. Either all or nothing.

pxlcoder commented on Source/WebCore/css/SelectorPseudoElementTypeMap.in:
```diff
@@ -22,3 +22,5 @@ placeholder, PseudoElementWebKitCustom
 -webkit-scrollbar-track-piece
 selection
 slotted
+thumb, PseudoElementWebKitCustom
+track, PseudoElementWebKitCustom
```
It's pretty strange to have an internal-only pseudo element that's unprefixed.

nt1m replied:
I think this is fine, the aim is to enable this in the future.

One thing that would be nice is to rename `PseudoElementWebKitCustom` to be more accurate though. It's being used for `::file-selector-button` / `::placeholder` / `::thumb` / `::track`, so it's WebKitCustom not necessarily accurate. But that can be a followup.

annevk replied:
https://bugs.webkit.org/show_bug.cgi?id=263976

pxlcoder commented on Source/WebCore/css/parser/CSSParserSelector.cpp:
```diff
@@ -50,8 +50,12 @@ std::unique_ptr<CSSParserSelector> CSSParserSelector::parsePagePseudoSelector(St
+    if (mode != UASheetMode) {
+        if (equalLettersIgnoringASCIICase(pseudoTypeString, "thumb"_s) || equalLettersIgnoringASCIICase(pseudoTypeString, "track"_s))
+            return nullptr;
+    }
```
This logic is likely going to run too frequently. After looking at this code more closely, my suggestion would be to:

1. Remove the changes here.
2. Give your pseudo element a `-webkit` prefix.
3. Update `CSSSelector::parsePseudoElementType` to return `PseudoElementUnknown` in the exceptional case, underneath `type == PseudoElementWebKitCustom`.

Then the logic wouldn't run for any non-custom pseudo elements.

nt1m replied:
I disagree we should add a `-webkit-` prefix. The end goal is to eventually expose this pseudo to the web. We should properly support shadow pseudo IDs without `-webkit-` prefixes.

pxlcoder replied:
Sounds good. We should still restrict the additional check behind a comparison to `PseudoElementWebKitCustom` though, to avoid unnecessary comparisons.

anttijk commented on Source/WebCore/html/shadow/SwitchThumbElement.cpp:
```diff
@@ -35,24 +37,16 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(SwitchThumbElement);
 
 Ref<SwitchThumbElement> SwitchThumbElement::create(Document& document)
 {
-    return adoptRef(*new SwitchThumbElement(document));
+    Ref element = adoptRef(*new SwitchThumbElement(document));
+    ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
+    element->setPseudo(ShadowPseudoIds::thumb());
+    return element;
+
 }
```
It seems to me that we don't need SwitchThumbElement at all. Can it be replaced with a plain HTMLDivElement?

We generally don't want to add any new custom element subclasses unless they are elements with actual semantics.

annevk replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=263968 to investigate this when we're closer to being done.

anttijk commented on Source/WebCore/html/shadow/SwitchTrackElement.cpp:
```diff
@@ -35,24 +37,15 @@ WTF_MAKE_ISO_ALLOCATED_IMPL(SwitchTrackElement);
 
 Ref<SwitchTrackElement> SwitchTrackElement::create(Document& document)
```
Same here.

anttijk commented on Source/WebCore/css/htmlSwitchControl.css:
```diff
@@ -25,5 +25,10 @@
+
+input[type="checkbox"][switch]::thumb, input[type="checkbox"][switch]::track {
+    appearance: auto;
+}
```
What are we trying to achieve here? Are we actually planning to expose these UA shadow tree elements to the web for arbitrary styling? I don't think there are many existing non-legacy examples of that (`::file-selector-button`?).

The modern way to style UA shadow is to have scoped style in the shadow tree. But maybe there are performance concerns?

annevk replied:
Yes, exposing these for arbitrary styling is the eventual goal.

annevk committed:
Add <input type=checkbox switch> internal pseudo-elements
https://bugs.webkit.org/show_bug.cgi?id=263928
rdar://117708673

Reviewed by Aditya Keerthi and Antti Koivisto.

Changing the internal appearance of internal shadow tree elements in
resolveCustomStyle() does not work completely as this is too late for
Adjuster::adjust() to consider it having an appearance for the purposes
of theming. In particular it does not allow us to set intrinsic sizes
for these elements.

As such, internal shadow tree elements need to have appearance:inherit
declared upon them upfront in a user agent style sheet.

As we are planning on eventually supporting pseudo-elements for the
track and thumb of switches, add those now, but only expose them
internally until we are sure they match the eventual standard. These
pseudo-elements can then be used to set appearance:inherit in the user
agent style sheet on the internal shadow tree elements.

This appearance value (when it computes to auto) means that
autoAppearanceForElement() is invoked at which point we can then map
them to the appropriate internal values for the track and thumb.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::parsePseudoElementType):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorPseudoElementTypeMap.in:
* Source/WebCore/css/htmlSwitchControl.css:
(input[type="checkbox"][switch]::thumb, input[type="checkbox"][switch]::track):
* Source/WebCore/css/parser/CSSParserSelector.cpp:
(WebCore::CSSParserSelector::parsePseudoElementSelector):
* Source/WebCore/css/parser/CSSParserSelector.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/html/shadow/ShadowPseudoIds.cpp:
(WebCore::ShadowPseudoIds::thumb):
(WebCore::ShadowPseudoIds::track):
* Source/WebCore/html/shadow/ShadowPseudoIds.h:
* Source/WebCore/html/shadow/SwitchThumbElement.cpp:
(WebCore::SwitchThumbElement::create):
(WebCore::SwitchThumbElement::resolveCustomStyle): Deleted.
* Source/WebCore/html/shadow/SwitchThumbElement.h:
* Source/WebCore/html/shadow/SwitchTrackElement.cpp:
(WebCore::SwitchTrackElement::create):
(WebCore::SwitchTrackElement::resolveCustomStyle): Deleted.
* Source/WebCore/html/shadow/SwitchTrackElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):

Canonical link: https://commits.webkit.org/270037@main

nt1m commented on Source/WebCore/css/htmlSwitchControl.css:
```diff
@@ -27,3 +27,7 @@
     margin: initial;
 }
+
+input[type="checkbox"][switch]::thumb, input[type="checkbox"][switch]::track {
+    appearance: inherit !important;
+}
```
Would have been nice to have tests for this

pxlcoder replied:
We can add ref tests once painting is implemented.
